<hr style="border:5px solid #002060"> </hr>

# pandas Fundamentals <hr style="border:4.5px solid #002060"> </hr>

## Introduction to pandas Series

Import the 'pandas' library to load it into the computer's memory, so that you can work with it in this Notebook Document.

<br/> *Note: Don't forget to use the widely-accepted convention as well.*

<br/> Remember that no matter how many times you execute this code cell, the library will be imported only once in this Document, and it will remain active.

In [3]:
#import pandas library
import pandas as pd

Create the **employee_names** list.

In [4]:
employee_names = ['Amy White', 'Jack Stewart', 'Richard Lauderdale', 'Sara Johnson']
employee_names

['Amy White', 'Jack Stewart', 'Richard Lauderdale', 'Sara Johnson']

Verify the **employee_names** object is a list.

In [5]:
type(employee_names)

list

"employee_names" is a list indeed.

Create a pandas Series object containing the elements from the **employee_names** list. Call it **employee_names_Series**.

In [6]:
employee_names_Series = pd.Series(employee_names)
employee_names_Series

0             Amy White
1          Jack Stewart
2    Richard Lauderdale
3          Sara Johnson
dtype: object

Confirm the object is of the Series type.

In [7]:
print(type(employee_names_Series))

<class 'pandas.core.series.Series'>


"employee_names_Series" is a pandas Series.

Now, create a Series object directly. That is, not from an existing list, but by using the following structure:
**pd.Series([...])**
<br/>Let the elements of the Series object be the following numbers: 5, 8, 3, and 10. Name the object **work_experience_yrs.**


In [8]:
work_experience_yrs = pd.Series([5,8,3,10])
work_experience_yrs

0     5
1     8
2     3
3    10
dtype: int64

Import the 'NumPy' package to load it into the computer's memory, so that you can work with it in this Notebook Document.

*Note: Don't forget to use the widely-accepted convention as well.*

In [9]:
#Import numpy library
import numpy as np

Execute the code cell below to create the **array_1** NumPy array object.

In [10]:
array_age = np.array([50, 53, 35, 43])
array_age

array([50, 53, 35, 43])

Verify the type of the **array_age** object:

In [11]:
print(type(array_age))

<class 'numpy.ndarray'>


Create a Series object called **series_age** from the NumPy array object **array_age** you just created.

In [12]:
series_age = pd.Series(array_age)
series_age

0    50
1    53
2    35
3    43
dtype: int64

Check the type of the newly created object.

In [13]:
print(type(series_age))

<class 'pandas.core.series.Series'>


Use the *print()* function to display the content of **series_age**.

In [14]:
print(series_age)

0    50
1    53
2    35
3    43
dtype: int64


## Working with Methods in Python

Consider the following Series object.

In [15]:
employees_work_exp = pd.Series({
'Amy White'   : 3,
'Jack Stewart'   : 5,
'Richard Lauderdale'  : 4.5,
'Sara Johnson'  : 22,
'Patrick Adams' : 28,
'Jessica Baker'  : 14,
'Peter Hunt'   : 4,
'Daniel Lloyd'  : 6,
'John Owen'   : 1.5,
'Jennifer Phillips'  : 10,
'Courtney Rogers'   : 4.5,
'Anne Robinson'  : 2,
})

In [16]:
employees_work_exp

Amy White              3.0
Jack Stewart           5.0
Richard Lauderdale     4.5
Sara Johnson          22.0
Patrick Adams         28.0
Jessica Baker         14.0
Peter Hunt             4.0
Daniel Lloyd           6.0
John Owen              1.5
Jennifer Phillips     10.0
Courtney Rogers        4.5
Anne Robinson          2.0
dtype: float64

Use a certain method to extract the top five values from this Series.

In [17]:
employees_work_exp.head()

Amy White              3.0
Jack Stewart           5.0
Richard Lauderdale     4.5
Sara Johnson          22.0
Patrick Adams         28.0
dtype: float64

Use another method to extract the last few rows of **employees_work_exp**.

In [18]:
employees_work_exp.tail()

Daniel Lloyd          6.0
John Owen             1.5
Jennifer Phillips    10.0
Courtney Rogers       4.5
Anne Robinson         2.0
dtype: float64

## Using .unique() and .nunique()

In [19]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Load the "Region.csv" file and use the .squeeze() method to convert it into a series. Store the information in a variable called **region_data**. Preview the data with the pandas *.head()* method.

In [20]:
region_data = pd.read_csv("/content/drive/MyDrive/Region.csv").squeeze()

Verify that **region_data** is a Series object.

In [21]:
print(type(region_data))

<class 'pandas.core.series.Series'>


In [22]:
region_data.head()

0    Region 2
1    Region 6
2    Region 3
3    Region 2
4    Region 3
Name: Region, dtype: object

Use the *.describe()* method to obtain descriptive statistics on the **region_data** Series.
<br>Think of how many unique values there are in the data set.
<br>*Please note that the statistics provided in the output exclude missing data.*

In [23]:
region_data.describe()

count         1042
unique          18
top       Region 6
freq           326
Name: Region, dtype: object

You can obtain some of the values from the previous output by using Python built-in functions or methods. Please extract the total number of values from the **region_data** Series, then the number of unique values. Finally, obtain a list containing all unique values from this Series.

In [24]:
print(region_data.describe()[0])

1042


<ipython-input-24-4dcd1e510422>:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(region_data.describe()[0])


In [25]:
region_data.value_counts().sum()

1042

There are tital 1042 values in region_data.

In [26]:
region_data.unique()

array(['Region 2', 'Region 6', 'Region 3', 'Region 1', 'Region 5',
       'Region 9', 'Region 7', 'Region 4', 'Region 12', 'Region 16',
       'Region 8', 'Region 10', 'Region 13', 'Region 15', 'Region 11',
       'Region 14', 'Region 17', 'Region 18'], dtype=object)

## Using .sort_values()

Sort the values without specifying any arguments.

*Please note that the numbers go from 1 to 18, but we are currently ordering the values as labels, not integers. 1 is succeeded by 10, then 11, and so on until 18. Then we have 2, 3, 4, etc., until 8, and eventually - 9.*

In [27]:
region_data.sort_values()

462    Region 1
347    Region 1
609    Region 1
610    Region 1
339    Region 1
         ...   
536    Region 9
450    Region 9
8      Region 9
842    Region 9
940    Region 9
Name: Region, Length: 1042, dtype: object

Sort the values, setting the *ascending* parameter equal to *True*.

In [28]:
region_data.sort_values(ascending = True)

462    Region 1
347    Region 1
609    Region 1
610    Region 1
339    Region 1
         ...   
536    Region 9
450    Region 9
8      Region 9
842    Region 9
940    Region 9
Name: Region, Length: 1042, dtype: object

Sort the values in **region_data** in descending order.

In [29]:
region_data.sort_values(ascending = False)

450    Region 9
842    Region 9
141    Region 9
940    Region 9
8      Region 9
         ...   
569    Region 1
892    Region 1
558    Region 1
557    Region 1
309    Region 1
Name: Region, Length: 1042, dtype: object

## Introduction to pandas DataFrames

Import the pandas and numpy libraries by using the relevant well-known conventions.

In [30]:
import pandas as pd
import numpy as np

The goal of this exercise is to solidify your knowledge and understanding of the fundamentals of using pandas DataFrames.

Execute the next code cell to create the array called **array_c**.

In [31]:
array_c = np.array([[1000, 50, 78.45], [700, 112, 35.5]])
array_c

array([[1000.  ,   50.  ,   78.45],
       [ 700.  ,  112.  ,   35.5 ]])

Convert **array_c** into a DataFrame, called **company_statistics**. Assign the column names and row labels from the following lists as column names and index of this DataFrame, respectively.

In [32]:
# column names
columns = ['No. of Employees', 'Avg. Annual Revenue ($)', 'Share price ($)']

In [33]:
# row labels
labels = ['Company A', 'Company B']

In [34]:
company_statistics = pd.DataFrame(array_c,columns = columns,index= labels)

Display the **company_statistics** DataFrame.

In [35]:
company_statistics

No. of Employees  Avg. Annual Revenue ($)  Share price ($)
Company A            1000.0                     50.0            78.45
Company B             700.0                    112.0            35.50

Import the *Sales-products.csv* dataset into a DataFrame object called **sales_products_data**.

*Note: Please use the default index provided for this DataFrame throughout this exercise.*

In [36]:
sales_products_data = pd.read_csv("/content/drive/MyDrive/Sales-products.csv")
sales_products_data

SaleID RetailerCountry  OrderMethod   RetailerType  \
0        SaleID_1   United States          Fax  Outdoors Shop   
1        SaleID_2   United States          Fax  Outdoors Shop   
2        SaleID_3   United States          Fax  Outdoors Shop   
3        SaleID_4   United States          Fax  Outdoors Shop   
4        SaleID_5   United States          NaN  Outdoors Shop   
...           ...             ...          ...            ...   
1271  SaleID_1272           Spain  Sales visit  Outdoors Shop   
1272  SaleID_1273           Spain  Sales visit  Outdoors Shop   
1273  SaleID_1274           Spain  Sales visit  Outdoors Shop   
1274  SaleID_1275           Spain  Sales visit  Outdoors Shop   
1275  SaleID_1276           Spain  Sales visit  Outdoors Shop   

               ProductLine           ProductType                    Product  \
0     Personal Accessories          Cooking Gear  TrailChef Deluxe Cook Set   
1         Sports Equipment          Cooking Gear     TrailChef Double Flame   
2                      NaN                 Tents                  Star Dome   
3                      NaN                 Tents               Star Gazer 2   
4     Personal Accessories             First Aid         Compact Relief Kit   
...                    ...                   ...                        ...   
1271  Personal Accessories                  Rope              Husky Rope 60   
1272    Outdoor Protection  Climbing Accessories      Firefly Climbing Lamp   
1273  Personal Accessories  Climbing Accessories            Firefly Charger   
1274  Personal Accessories                 Tools                Granite Axe   
1275     Camping Equipment                 Tools            Granite Extreme   

        Year  Quarter   Revenue  Quantity  GrossMargin  
0     2018.0  Q1 2018  59728.66     491.0     0.357548  
1     2018.0  Q1 2018  36050.32     254.0     0.484274  
2     2018.0  Q1 2018  90040.48     149.0          NaN  
3     2018.0  Q1 2018       NaN     305.0     0.292938  
4     2018.0  Q1 2018       NaN     182.0     0.610710  
...      ...      ...       ...       ...          ...  
1271  2020.0  Q3 2020  30916.50     173.0     0.298114  
1272  2020.0  Q3 2020   7536.29     193.0     0.445287  
1273  2020.0  Q3 2020  12306.48     238.0     0.568420  
1274  2020.0  Q3 2020  56499.00    1472.0     0.490667  
1275  2020.0  Q3 2020  89427.00    1178.0     0.386895  

[1276 rows x 12 columns]

Verify that the obtained object is a pandas DataFrame.

In [37]:
print(type(sales_products_data))

<class 'pandas.core.frame.DataFrame'>


Try out a well-known pandas method used for Series on this DataFrame to obtain its last five rows.

In [38]:
sales_products_data.tail()

SaleID RetailerCountry  OrderMethod   RetailerType  \
1271  SaleID_1272           Spain  Sales visit  Outdoors Shop   
1272  SaleID_1273           Spain  Sales visit  Outdoors Shop   
1273  SaleID_1274           Spain  Sales visit  Outdoors Shop   
1274  SaleID_1275           Spain  Sales visit  Outdoors Shop   
1275  SaleID_1276           Spain  Sales visit  Outdoors Shop   

               ProductLine           ProductType                Product  \
1271  Personal Accessories                  Rope          Husky Rope 60   
1272    Outdoor Protection  Climbing Accessories  Firefly Climbing Lamp   
1273  Personal Accessories  Climbing Accessories        Firefly Charger   
1274  Personal Accessories                 Tools            Granite Axe   
1275     Camping Equipment                 Tools        Granite Extreme   

        Year  Quarter   Revenue  Quantity  GrossMargin  
1271  2020.0  Q3 2020  30916.50     173.0     0.298114  
1272  2020.0  Q3 2020   7536.29     193.0     0.445287  
1273  2020.0  Q3 2020  12306.48     238.0     0.568420  
1274  2020.0  Q3 2020  56499.00    1472.0     0.490667  
1275  2020.0  Q3 2020  89427.00    1178.0     0.386895

## pandas DataFrames - Common Attributes

Import the *Sales-products.csv* dataset into a DataFrame object called **sales_products_data**.

*Note: Please use the default index provided for this DataFrame throughout this exercise.*

Return the index of the **sales_products_data** DataFrame.

In [39]:
sales_products_data.index

RangeIndex(start=0, stop=1276, step=1)

Verify that **sales_products_data** is a RangeIndex object in an alternative way by using the **type()** funciton.

In [40]:
print(type(sales_products_data.index))

<class 'pandas.core.indexes.range.RangeIndex'>


Retrieve and Index object containing all column names from the **sales_products_data** DataFrame.

In [41]:
columns = sales_products_data.columns
columns

Index(['SaleID', 'RetailerCountry', 'OrderMethod', 'RetailerType',
       'ProductLine', 'ProductType', 'Product', 'Year', 'Quarter', 'Revenue',
       'Quantity', 'GrossMargin'],
      dtype='object')

You are requested to create a new DataFrame, called **df**, whose data is obtained from the the names of the columns of **sales_products_data** DataFrame, while the column names are obtained from the **new_column_names** list.
<br/><br/> First, execute the next cell to create **new_column_names**, and then in the next cell try creating the **df** DataFrame yourself (by helping yourself with the newly created **new_column_names** list).
<br/><br/> *Note: There are two ways to achieve this - one would be through utilizing the* **columns** *parameter, and the other – with the help of the* **"columns"** *attribute. Choose whichever option you prefer (or why not try both!).*

In [42]:
new_column_names = ['SaleID', 'Country', 'OrderMethod', 'Retailer', 'Line',
       'Type', 'Product', 'YearOfSales', 'Quarter', 'Revenue', 'Quantity',
       'Margin']

In [45]:

df = pd.DataFrame(data = [sales_products_data.columns],columns = new_column_names)

In [46]:
df

SaleID          Country  OrderMethod      Retailer         Line  \
0  SaleID  RetailerCountry  OrderMethod  RetailerType  ProductLine   

          Type  Product YearOfSales  Quarter  Revenue  Quantity       Margin  
0  ProductType  Product        Year  Quarter  Revenue  Quantity  GrossMargin

Display the "head" of the **df** DataFrame.

In [47]:
df.head()

SaleID          Country  OrderMethod      Retailer         Line  \
0  SaleID  RetailerCountry  OrderMethod  RetailerType  ProductLine   

          Type  Product YearOfSales  Quarter  Revenue  Quantity       Margin  
0  ProductType  Product        Year  Quarter  Revenue  Quantity  GrossMargin

Use an attribute to obtain metadata about the types of data stored in the different columns of the **sales_products_data** DataFrame.

In [48]:
df.dtypes

SaleID         object
Country        object
OrderMethod    object
Retailer       object
Line           object
Type           object
Product        object
YearOfSales    object
Quarter        object
Revenue        object
Quantity       object
Margin         object
dtype: object

Obtain a Numpy array containing all values of the **sales_products_data**.
<br/><br/> *Note: There are two ways to achieve this! Try both, but keep in mind that pandas developers advise us to stick to using the method.*

In [50]:
np_array = np.array(sales_products_data)

In [51]:
np_array

array([['SaleID_1', 'United States', 'Fax', ..., 59728.66, 491.0,
        0.35754797],
       ['SaleID_2', 'United States', 'Fax', ..., 36050.32, 254.0,
        0.4842745],
       ['SaleID_3', 'United States', 'Fax', ..., 90040.48, 149.0, nan],
       ...,
       ['SaleID_1274', 'Spain', 'Sales visit', ..., 12306.48, 238.0,
        0.56842037],
       ['SaleID_1275', 'Spain', 'Sales visit', ..., 56499.0, 1472.0,
        0.49066667],
       ['SaleID_1276', 'Spain', 'Sales visit', ..., 89427.0, 1178.0,
        0.38689474]], dtype=object)

Verify that the obtained objects are NumPy arrays with the help of the "Type" function.

In [52]:
type(np_array)

numpy.ndarray

In [ ]:
#Import the pandas and numpy libraries by using the relevant well-known conventions.

import pandas as pd
import numpy as np

Obtain the number of rows and columns in the **sales_products_data** DataFrame.

In [ ]:
sales_products_data.shape #shape of sales_products_data

(1276, 12)

Check if the output obtained in the previous cell is a tuple.

In [ ]:
type(sales_products_data.shape) #type of shape

tuple

## Data Selection in pandas DataFrames

Although we are not advised to, retrieve the data from the 'Product' column from the **sales_products_data** column *without* using the indexing operator **[ ]**.

In [ ]:
sales_products_data.Product #fetching "Product" data

0       TrailChef Deluxe Cook Set
1          TrailChef Double Flame
2                       Star Dome
3                    Star Gazer 2
4              Compact Relief Kit
                  ...            
1271                Husky Rope 60
1272        Firefly Climbing Lamp
1273              Firefly Charger
1274                  Granite Axe
1275              Granite Extreme
Name: Product, Length: 1276, dtype: object

Retrieve the data from the 'Product' column from the **sales_products_data** column by using the indexing operator **[ ]**.

In [ ]:
sales_products_data["Product"]

0       TrailChef Deluxe Cook Set
1          TrailChef Double Flame
2                       Star Dome
3                    Star Gazer 2
4              Compact Relief Kit
                  ...            
1271                Husky Rope 60
1272        Firefly Climbing Lamp
1273              Firefly Charger
1274                  Granite Axe
1275              Granite Extreme
Name: Product, Length: 1276, dtype: object

Apply the **.tail()** method to the code you used in your previous response to obtain the last five rows from the "Product" column.

In [ ]:
sales_products_data.Product.tail()

1271            Husky Rope 60
1272    Firefly Climbing Lamp
1273          Firefly Charger
1274              Granite Axe
1275          Granite Extreme
Name: Product, dtype: object

Although we are not advised to, retrieve the data from the 'Quarter' column from the **sales_products_data** column *without* using the indexing operator **.[ ]**.

In [ ]:
sales_products_data.Quarter

0       Q1 2018
1       Q1 2018
2       Q1 2018
3       Q1 2018
4       Q1 2018
         ...   
1271    Q3 2020
1272    Q3 2020
1273    Q3 2020
1274    Q3 2020
1275    Q3 2020
Name: Quarter, Length: 1276, dtype: object

Retrieve the data from the 'Quarter' column from the **sales_products_data** column by using the indexing operator **[ ]**.

In [ ]:
sales_products_data["Quarter"]

0       Q1 2018
1       Q1 2018
2       Q1 2018
3       Q1 2018
4       Q1 2018
         ...   
1271    Q3 2020
1272    Q3 2020
1273    Q3 2020
1274    Q3 2020
1275    Q3 2020
Name: Quarter, Length: 1276, dtype: object

Use any of the 4 previous examples to verify that the data obtained from a column of a DataFrame is actually stored in a Series object:

In [ ]:
print(type(sales_products_data["Quarter"]))

<class 'pandas.core.series.Series'>


<br> Use a well-know attribute and attribute chaining to obtain information about the type of data stored in the "Revenue" column.

In [ ]:
sales_products_data.Revenue.dtype

dtype('float64')

In [ ]:
sales_products_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1276 entries, 0 to 1275
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   SaleID           1276 non-null   object 
 1   RetailerCountry  1276 non-null   object 
 2   OrderMethod      1271 non-null   object 
 3   RetailerType     1273 non-null   object 
 4   ProductLine      1273 non-null   object 
 5   ProductType      1273 non-null   object 
 6   Product          1272 non-null   object 
 7   Year             1275 non-null   float64
 8   Quarter          1276 non-null   object 
 9   Revenue          1260 non-null   float64
 10  Quantity         1273 non-null   float64
 11  GrossMargin      1250 non-null   float64
dtypes: float64(4), object(8)
memory usage: 119.8+ KB


**Selecting multiple columns from a DF**

Use a list with column names within the indexing operator ([ ]) to extract data from the 'Revenue' and 'Quantity' columns from the **sales_products_data** DataFrame. Use the **.head()** method to only display the top five rows of the selection.  <br>
Solve the problem by using a single line of code.

In [ ]:
sales_products_data[["Revenue","Quantity"]].head()

Revenue  Quantity
0  59728.66     491.0
1  36050.32     254.0
2  90040.48     149.0
3       NaN     305.0
4       NaN     182.0

Can you think of a more elegant way to obtain the same result? This option is more often encountered in professional work since it improves the legibility of your code.

In [ ]:
sales_products_data.loc[:,["Revenue","Quantity"]].head()

Revenue  Quantity
0  59728.66     491.0
1  36050.32     254.0
2  90040.48     149.0
3       NaN     305.0
4       NaN     182.0

This could be a method but first one seems more better.

## pandas DataFrames - Indexing with .iloc[]

Select the data from the first row from the **sales_products_data** DataFrame (don't forget that in programming we start counting from 0!).

In [ ]:
sales_products_data.iloc[0:1]

SaleID RetailerCountry OrderMethod   RetailerType           ProductLine  \
0  SaleID_1   United States         Fax  Outdoors Shop  Personal Accessories   

    ProductType                    Product    Year  Quarter   Revenue  \
0  Cooking Gear  TrailChef Deluxe Cook Set  2018.0  Q1 2018  59728.66   

   Quantity  GrossMargin  
0     491.0     0.357548

Select the data from the penultimate row from the **sales_products_data** DataFrame.

In [ ]:
sales_products_data.iloc[1274:1275]

SaleID RetailerCountry  OrderMethod   RetailerType  \
1274  SaleID_1275           Spain  Sales visit  Outdoors Shop   

               ProductLine ProductType      Product    Year  Quarter  Revenue  \
1274  Personal Accessories       Tools  Granite Axe  2020.0  Q3 2020  56499.0   

      Quantity  GrossMargin  
1274    1472.0     0.490667

Select the value from the second row and sixth column of **sales_products_data** by using the **.iloc[]** accessor.
<br><br> *Hint: If you've worked correctly, you should obtain 'Cooking Gear'.*

In [ ]:
sales_products_data.iloc[0,5]

'Cooking Gear'

Use the **.shape** attribute to discover the largest numbers that you can use as arguments of the **.iloc[]** accessor.

In [ ]:
sales_products_data.shape

(1276, 12)

We can use [1275,11] as the largest number to be used as an argument for iloc[].

Considering the values you just obtained, use the **.iloc[]** accessor find out the value that has been stored in the last row and farthest-to-the-right column.

Last row is 1275 and farthest-to-the-right column is "GrossMargin" i.e., column 12(index 11)

In [ ]:
sales_products_data.iloc[1275,11]

0.38689474

Verify that you've obtained the correct value by applying the **.tail()** method to the **sales_products_data** DataFrame.
<br> Observe the value that is farthest to the right on the last row.

In [ ]:
sales_products_data.tail()

SaleID RetailerCountry  OrderMethod   RetailerType  \
1271  SaleID_1272           Spain  Sales visit  Outdoors Shop   
1272  SaleID_1273           Spain  Sales visit  Outdoors Shop   
1273  SaleID_1274           Spain  Sales visit  Outdoors Shop   
1274  SaleID_1275           Spain  Sales visit  Outdoors Shop   
1275  SaleID_1276           Spain  Sales visit  Outdoors Shop   

               ProductLine           ProductType                Product  \
1271  Personal Accessories                  Rope          Husky Rope 60   
1272    Outdoor Protection  Climbing Accessories  Firefly Climbing Lamp   
1273  Personal Accessories  Climbing Accessories        Firefly Charger   
1274  Personal Accessories                 Tools            Granite Axe   
1275     Camping Equipment                 Tools        Granite Extreme   

        Year  Quarter   Revenue  Quantity  GrossMargin  
1271  2020.0  Q3 2020  30916.50     173.0     0.298114  
1272  2020.0  Q3 2020   7536.29     193.0     0.445287  
1273  2020.0  Q3 2020  12306.48     238.0     0.568420  
1274  2020.0  Q3 2020  56499.00    1472.0     0.490667  
1275  2020.0  Q3 2020  89427.00    1178.0     0.386895

The value I get is 0.38689474 but the tail() function is giving the rounded version i.e., 0.386895,but it's the same.

Prove that **.iloc[]** can be applied to a Series object such as **sales_products_data['Product']** as well (e.g. by obtaining the first element from this column).
<br><br> *Note: We are asking you to try this syntax out to make sure you are in control of obtaining the desired portion of the data regardless of the object it has been contained into. The syntax and tools are similar, while their utilization may differ.*
<br> *This note applies to this and the next task.*

In [ ]:
sales_products_data['Product'].iloc[0]

'TrailChef Deluxe Cook Set'

In [ ]:
sales_products_data.head(1)

SaleID RetailerCountry OrderMethod   RetailerType           ProductLine  \
0  SaleID_1   United States         Fax  Outdoors Shop  Personal Accessories   

    ProductType                    Product    Year  Quarter   Revenue  \
0  Cooking Gear  TrailChef Deluxe Cook Set  2018.0  Q1 2018  59728.66   

   Quantity  GrossMargin  
0     491.0     0.357548

As we can see iloc[] successfully accessed first element from the "Product" feature.So it can be applied to series as well.

Although we are not advised to, when working with a Series object, we can obtain the same output without using the **.iloc[]** indexer and using "classical" Python indexing. Can you try that out?

In [ ]:
sales_products_data["Product"][0]

'TrailChef Deluxe Cook Set'

## pandas DataFrames - Indexing with .loc[]

Retrieve the values from the first row of the DataFrame by using the specified label-based index rather than the default numeric index.

In [ ]:
sales_products_data.loc[0:0]

SaleID RetailerCountry OrderMethod   RetailerType           ProductLine  \
0  SaleID_1   United States         Fax  Outdoors Shop  Personal Accessories   

    ProductType                    Product    Year  Quarter   Revenue  \
0  Cooking Gear  TrailChef Deluxe Cook Set  2018.0  Q1 2018  59728.66   

   Quantity  GrossMargin  
0     491.0     0.357548

By using the specified label-based index again, retrieve the information stored in the 'GrossMargin' column for the 4th sale.

In [ ]:
sales_products_data["GrossMargin"].loc[3]

0.29293788

Obtain a DataFrame containg the values stored in all columns for sales numbered 15, 235, and 1147 by using the **.loc[]** indexer.

In [ ]:
sales_products_data.loc[[15,235,1147]]

SaleID RetailerCountry OrderMethod      RetailerType  \
15      SaleID_16   United States         Web              Mall   
235    SaleID_236          Mexico         Web              Mall   
1147  SaleID_1148         Germany       Phone  Department Store   

            ProductLine ProductType        Product    Year  Quarter   Revenue  \
15    Camping Equipment      Knives      Max Gizmo  2018.0  Q1 2018  12362.20   
235   Camping Equipment  Binoculars  Ranger Vision  2018.0  Q3 2018   5700.00   
1147  Camping Equipment    Lanterns  EverGlow Lamp  2020.0  Q1 2020  44998.55   

      Quantity  GrossMargin  
15       323.0     0.549791  
235       37.0     0.370000  
1147    1670.0     0.524364

With the help of the **.loc[]** indexer, select the last 5 records from the 'Year' column.

In [ ]:
sales_products_data["Year"].loc[1271:1275]

1271    2020.0
1272    2020.0
1273    2020.0
1274    2020.0
1275    2020.0
Name: Year, dtype: float64